### PR TITLE
Lists: Undo Exceptions Mark II

### PR DESCRIPTION
**Pull request by jleandroperez**
### Fix
In this PR we're reimplementing the mechanism(s) that deal with Tabs and Newlines:

- Newlines now insert a **SPTextAttachment directly** when it's needed (fixes UndoManager crashes)
- Tabs are now allowed anywhere within a line that contains (any kind) of bullet: The whole line is indented!
- Unit tests for everything and anything!

@aerych @diegoreymendez gentlemen, may I bug you with a really fun PR?

I've split the test scenarios in two, so that it's not massive for anyone.

**Thank you both in advance!!!**

Ref. #467

---

@aerych Whenever you have a few, mind taking a peek at these? 👇 

### Test: Tabs on Plain Text
1. Add random text on a new note
2. Place the cursor anywhere
3. Press **Tab**

- [x] Verify the plain text is properly indented

### Test: Tabs on Selected Text
1. Add random text on a new note
2. Select Everything
3. Press **Tab**

- [x] Verify the entire text is replaced with a tab!

### Test: Tabs on Lists
1. Insert a list (**Format** > **Insert Checklist**)
2. Enter few items

- [x] Verify that pressing `tab` anywhere on any line indents such line
- [x] Verify that selecting text, and pressing `tab` indents **only** the first line

**Note:** we might revisit the selected-text-indented behavior. Preserving the current `develop` behavior for simplicity.

---

@diegoreymendez Can I bother you with these scenarios? 👇 

### Test: Newline on Empty List Item
1. Insert a list (**Format** > **Insert Checklist**)
2. Hit **newline** on the empty list item

- [x] Verify the list marker is removed

### Test: Cursor Before List Item
1. Insert a list (**Format** > **Insert Checklist**)
2. Place the cursor on the left hand side of the list marker
3. Press newline

- [x] Verify the checklist is moved to the new line
- [x] Verify that we don't end up with two checklist markers on the same line

### Test: Newline on List Item
1. Insert a list (**Format** > **Insert Checklist**)
2. Place the cursor before the list marker, and add few spaces
3. Place the cursor on the right hand side of the marker
4. Insert some random text
5. Hit **newline**

- [x] Verify a new line is added
- [x] Verify that the new line gets the exact same spaces you've added in (2)

---

@diegoreymendez and @aerych Can you both give these a shot please?👇 

1. Add a new note
2. Enter some random title
3. Insert a list (**Format** > **Insert Checklist**)
4. Type some random text for the first item
5. Edit the Title: add a character
6. Click over the list item you've added in (4): hit newline and add new text
6. Repeat steps 4 thru 6 a few times

- [x] Verify that hitting CMD + Z does not break the Undo Manager (no errors in the console!)

**Note:** The Undo Manager seems to be "saving" checkpoints whenever the "last edited spots" aren't contiguous. That's why we need to edit the TITLE > Item (and loop!).

**Thank you gentlemen!!!**

### Release
These changes do not require release notes.

### CODE REVIEW

**Comment by aerych** [comment]
Hola @jleandroperez.  When testing the undo behavior I saw something weird.  The first time I tried, it seemed like undo was working but once I'd undo'd back to an empty note I started seeing exceptions. I wasn't able to reproduce the exception on subsequent tries. So take it as an FYI. I'll include a copy below.
I stopped and restarted the simulator and tried to reproduce the error in the console, and ended up finding something different.  If I hold down control Z to erase everything I've added to a new note, I'm left with a single checkbox for a list. If I repeat the test, adding another title, more content, more lists, etc. then undo everything again, I ended up with a checkmark, and maybe some random text, and maybe another checkmark that the undo command won't clear. If I manually delete it then tap undo it comes back.  Repeat the test multiple times, undoing everything each time and the "left over junk" just keeps growing.  I'll send a recording to demo it.

> 2020-03-20 15:59:24.160178-0500 Simplenote[44786:1063538] [General] undo: NSUndoManager 0x6000021587d0 is in invalid state, undo was called with too many nested undo groups
> 2020-03-20 15:59:24.161779-0500 Simplenote[44786:1063538] [General] (
> 	0   CoreFoundation                      0x00007fff34b3c8ab __exceptionPreprocess + 250
> 	1   libobjc.A.dylib                     0x00007fff6adf6805 objc_exception_throw + 48
> 	2   Foundation                          0x00007fff3725b6de -[NSUndoManager undoNestedGroup] + 0
> 	3   AppKit                              0x00007fff31ec8d38 -[NSApplication(NSResponder) sendAction:to:from:] + 299
> 	4   AppKit                              0x00007fff31fd6c82 -[NSMenuItem _corePerformAction] + 312
> 	5   AppKit                              0x00007fff31fd69fe -[NSCarbonMenuImpl performActionWithHighlightingForItemAtIndex:] + 106
> 	6   AppKit                              0x00007fff31fd59a0 -[NSMenu performKeyEquivalent:] + 348
> 	7   AppKit                              0x00007fff32409a4b routeKeyEquivalent + 521
> 	8   AppKit                              0x00007fff31e2964c -[NSApplication(NSEvent) sendEvent:] + 1072
> 	9   Simplenote                          0x00000001000460c5 -[SPApplication sendEvent:] + 341
> 	10  AppKit                              0x00007fff31c750cf -[NSApplication run] + 707
> 	11  AppKit                              0x00007fff31c47465 NSApplicationMain + 777
> 	12  Simplenote                          0x0000000100005352 main + 34
> 	13  libdyld.dylib                       0x00007fff6c1647fd start + 1
> )

File: Simplenote/NSString+Simplenote.swift [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    ///
+    func line(at location: Int) -> (NSRange, String) {
+        let queryRange = NSRange(location: location, length: .zero)
+        let resultRange = lineRange(for: queryRange)
+        let resultText = substring(with: resultRange)
+
+        return (resultRange, resultText)
+    }
+
+    /// Returns the Range of the List Marker: This also includes NSTextAttachments
```
Comment (aerych) [inline Simplenote/NSString+Simplenote.swift]: What does "This also includes NSTextAttachments" mean?  Is it referring to the NSTextAttachment that is part of the .listMarkers array?

File: Simplenote/NSString+Simplenote.swift [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        for prefix in prefixes where trimmedString.hasPrefix(prefix) {
+            return range(of: prefix)
+        }
+
+        return nil
+    }
+
+    /// Returns the UnicodeScalar at the specified location, if any.
+    ///
+    /// - Note: For convenience, we'll check bounds!
```
Comment (aerych) [inline Simplenote/NSString+Simplenote.swift]: `we'll check bounds` ?  Is this referring to checking that the passed location is less than the length of the NSString?

File: SimplenoteTests/NSTextViewSimplenoteTests.swift [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+            "\t+ ",
+            "  + ",
+            marker,
+            "\t\t\t" + marker
+        ]
+    }
+
+    var samplePlainText: [String] {
+        return [
+            "Here comes the sun, here comes the sun\n",
```
Comment (aerych) [inline SimplenoteTests/NSTextViewSimplenoteTests.swift]: Maybe we shouldn't use copyrighted song lyrics here :)

Comment (jleandroperez) [reply SimplenoteTests/NSTextViewSimplenoteTests.swift]: You know... I thought it was already public domain, but huge nope!! Updating, thanks for the heads up!!

File: Simplenote/NSString+Simplenote.swift (hunk 2 above)
Comment (jleandroperez) [reply Simplenote/NSString+Simplenote.swift]: Updating to something a bit more clear! Thank you sir!

```
For convenience, we'll disregard any `location` that is greater than the receiver's length
```

File: Simplenote/NSString+Simplenote.swift (hunk 1 above)
Comment (jleandroperez) [reply Simplenote/NSString+Simplenote.swift]: Ahhhh forgot to patch that. Old code did not support indentation of Lists that had a TextAttachment (that is, only lists with a text based bullet were being indented).

In this PR that was changed, and we just support all the things. Patching up!

**Comment by jleandroperez** [comment]
@aerych I **think** the pending exceptions are related to the way (new) checklists are inserted.

There are (Still) two spots that tamper with the TextStorage directly, way beyond the TextView's reach. 

Mind if I address those in a follow up PR? (the size of this one is already super huge!).

**Thank you!!!**

**Comment by jleandroperez** [comment]
For the record, there are one (or two) known issues, to be addressed in follow ups.

@diegoreymendez @aerych :hug: Thank you both gentlemen!!!

**Pull request merged**